config/decode.go

```diff
@@ -0,0 +1,45 @@
+package config
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+)
+
+func processUnusedConfigKeys(md mapstructure.Metadata, file string) error {
+	if len(md.Unused) == 0 {
+		return nil
+	}
+
+	sort.Strings(md.Unused)
+	err := fmt.Errorf("'%s' has invalid keys: %s", file, strings.Join(md.Unused, ", "))
+
+	for _, key := range md.Unused {
+		if key == "provider" {
+			err = fmt.Errorf(`%s
+	'provider' is an invalid key for Consul Terraform Sync configuration, try 'terraform_provider'.
+	terraform_provider configuration blocks are similar to provider blocks in Terraform but have additional features supported only by Consul Terraform Sync.
+
+`, err)
+		}
+	}
+	return err
+}
+
+// decodeError is a middleware for mapstructure.Decoder.Decode() errors
+// with decode hooks specifically for CTS.
+func decodeError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	r := regexp.MustCompile(`'([^\']+)' expected a map, got 'slice'`)
```

Comment:
I usually see regex as a global variable because it's safer since `regex.MustCompile` can panic. Is there maybe a use-case for declaring the regex locally?

Reply:
oh that's true about the panic, vs global it'll panic earlier. It was an oversight 🙂